test/end2end_test.go

```diff
@@ -5509,6 +5510,131 @@ func (s) TestClientForwardsGrpcAcceptEncodingHeader(t *testing.T) {
 	}
 }
 
+func (s) TestServerHandlerSetSendCompressor(t *testing.T) {
+	for _, tt := range []struct {
+		desc          string
+		resCompressor string
+		wantErr       error
+	}{
+		{
+			desc:          "gzip_response_compressor",
+			resCompressor: "gzip",
+		},
+		{
+			desc:          "identity_response_compressor",
+			resCompressor: "identity",
+		},
+		{
+			desc:          "unregistered_snappy2_response_compressor",
+			resCompressor: "snappy2",
+			wantErr:       fmt.Errorf("rpc error: code = Internal desc = grpc: unable to set send compressor: compressor not registered \"snappy2\""),
+		},
+	} {
+		t.Run(tt.desc, func(t *testing.T) {
+			testSetSendCompressorHandler(t, tt.resCompressor, tt.wantErr)
+		})
+	}
+}
+
+func (s) TestUnadvertisedSendCompressorFailure(t *testing.T) {
+	// disable client compressor advertisement.
+	defer func(b bool) { envconfig.AdvertiseCompressors = b }(envconfig.AdvertiseCompressors)
+	envconfig.AdvertiseCompressors = false
+
+	wantErr := fmt.Errorf("rpc error: code = Internal desc = grpc: unable to set send compressor: client does not support compressor \"gzip\"")
+	testSetSendCompressorHandler(t, "gzip", wantErr)
+}
+
+func testSetSendCompressorHandler(t *testing.T, resCompressor string, wantErr error) {
+	ss := &stubserver.StubServer{
+		EmptyCallF: func(ctx context.Context, in *testpb.Empty) (*testpb.Empty, error) {
+			if err := grpc.SetSendCompressor(ctx, resCompressor); err != nil {
+				return nil, err
+			}
+			return &testpb.Empty{}, nil
+		},
+		FullDuplexCallF: func(stream testpb.TestService_FullDuplexCallServer) error {
+			if _, err := stream.Recv(); err != nil {
+				return err
+			}
+
+			if err := grpc.SetSendCompressor(stream.Context(), resCompressor); err != nil {
+				return err
+			}
+
+			return stream.Send(&testpb.StreamingOutputCallResponse{})
+		},
+	}
+	if err := ss.Start(nil); err != nil {
+		t.Fatalf("Error starting endpoint server: %v", err)
+	}
+	defer ss.Stop()
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	_, err := ss.Client.EmptyCall(ctx, &testpb.Empty{}, grpc.UseCompressor("gzip"))
+	if fmt.Sprint(err) != fmt.Sprint(wantErr) {
+		t.Fatalf("Unexpected unary call error, got: %v, want: %v", err, wantErr)
+	}
+
+	s, err := ss.Client.FullDuplexCall(ctx)
+	if err != nil {
+		t.Fatalf("Unexpected full duplex call error, got: %v, want: nil", err)
+	}
+
+	if err := s.Send(&testpb.StreamingOutputCallRequest{}); err != nil {
+		t.Fatalf("Unexpected full duplex call send error, got: %v, want: nil", err)
+	}
+
+	if _, err := s.Recv(); fmt.Sprint(err) != fmt.Sprint(wantErr) {
+		t.Fatalf("Unexpected full duplex recv error, got: %v, want: %v", err, wantErr)
+	}
+}
+
+func (s) TestSendCompressorFailureAfterHeaderSend(t *testing.T) {
```

Comment:
I would be totally OK if you split up the tests to have separate tests for unary and streaming RPCs instead of doing them in the same test. Thanks.

Reply:
done